src/dac.rs

```diff
@@ -0,0 +1,248 @@
+//! Configure the internal DAC on the stm32f3xx.
+//! Incomplete, but includes basic operation.
+//!
+//! You should have the appropriate DAC pin set up as an analog input, to prevent
+//! parasitic power consumption. For example:
+//! ```rust
+//! let _dac1_pin = gpioa.pa4.into_analog(&mut gpioa.moder, &mut gpioa.pupdr);
+//! let _dac2_pin = gpioa.pa5.into_analog(&mut gpioa.moder, &mut gpioa.pupdr);
+//! ```
+
+use core::{convert::Infallible, fmt};
+
+use crate::{
+    gpio::{
+        gpioa::{PA4, PA5},
+        Analog,
+    },
+    pac,
+    rcc::APB1,
+};
+
+/// todo: DAC trait that will go in a PR to `embeddd-hal`.
+/// Once merged and released in that crate, remove from this module.
+/// [PR](https://github.com/rust-embedded/embedded-hal/pull/247)
+pub trait SingleChannelDac {
+    /// Error type returned by DAC methods
+    type Error;
+    type Word;
+
+    /// Output a constant signal, given a bit word.
+    fn try_set_value(&mut self, value: Self::Word) -> Result<(), Self::Error>;
+}
+
+/// This is an abstraction to ensure that the DAC output pin is configured
+/// as an analog output.
+pub trait Pins {}
+impl Pins for PA4<Analog> {}
+impl Pins for PA5<Analog> {}
+impl Pins for (PA4<Analog>, PA5<Analog>) {}
```

Comment:
`PA4` is for channel 1 and `PA5` for channel 2, right? Your DAC constructor only allows selecting a single channel at a time. So why would one want to pass in both pins?